src/elysium/sigmawalletv1.cpp

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2020 The Zcoin Core Developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "sigmawalletv1.h"
+
+#include "../libzerocoin/Zerocoin.h"
+#include "../sigma/openssl_context.h"
+#include "../wallet/wallet.h"
+
+namespace elysium {
+
+SigmaWalletV1::SigmaWalletV1()
+    : SigmaWallet(new SigmaWalletV1::Database())
+{
+}
+
+bool SigmaWalletV1::GeneratePublicKey(ECDSAPrivateKey const &priv, secp256k1_pubkey &out)
+{
+    return secp256k1_ec_pubkey_create(
+        OpenSSLContext::get_context(),
+        &out,
+        priv.data());
+}
+
+void SigmaWalletV1::GenerateSerial(secp256k1_pubkey const &pubkey, secp_primitives::Scalar &serial)
+{
+    std::array<uint8_t, 33> compressedPub;
+
+    size_t outSize = sizeof(compressedPub);
+    secp256k1_ec_pubkey_serialize(
+        OpenSSLContext::get_context(),
+        compressedPub.begin(),
+        &outSize,
+        &pubkey,
+        SECP256K1_EC_COMPRESSED);
+
+    if (outSize != 33) {
+        throw std::runtime_error("Compressed public key size is invalid.");
+    }
+
+    std::array<uint8_t, CSHA256::OUTPUT_SIZE> hash;
+    CSHA256().Write(compressedPub.begin(), sizeof(compressedPub)).Finalize(hash.data());
+
+    serial.memberFromSeed(hash.begin());
+}
+
+uint32_t SigmaWalletV1::BIP44ChangeIndex() const
+{
+    return BIP44_ELYSIUM_MINT_INDEX_V1;
+}
+
+SigmaPrivateKey SigmaWalletV1::GeneratePrivateKey(uint512 const &seed)
+{
+    std::array<uint8_t, 32> ecdsaKeyOut;
+    return GeneratePrivateKey(seed, ecdsaKeyOut);
+}
+
+SigmaPrivateKey SigmaWalletV1::GeneratePrivateKey(
+    uint512 const &seed, ECDSAPrivateKey &ecdsaKeyOut)
+{
+    // first 32 bytes as seed of ecdsa key and serial
+    std::array<uint8_t, 32> tmp;
+    std::copy(seed.begin(), seed.begin() + 32, ecdsaKeyOut.begin());
+
+    // hash until get valid private key
+    secp256k1_pubkey pubkey;
+    do {
+        std::copy(ecdsaKeyOut.begin(), ecdsaKeyOut.end(), tmp.begin());
+        CSHA256().Write(tmp.begin(), tmp.size()).Finalize(ecdsaKeyOut.begin());
```

Comment:
Better to use `tmp.data()` instead of `tmp.begin()` so it matched with the purpose.